File: bench/benches/sortby.lua
-- sortby: N (key, idx) records ordered by a user comparator on a derived key (key%4096, then key);
-- checksum = rolling hash of the idx sequence.
package.path = (arg[0]:match("(.*/)") or "./") .. "../lib/?.lua;" .. package.path
local bench = require("bench")
local N = 5000
bench("sortby", function()
  local x = 1
  local data = {}
  for i = 1, N do x = (16807 * x) % 2147483647; data[i] = {x, i - 1} end
  table.sort(data, function(a, b)
    local ma, mb = a[1] % 4096, b[1] % 4096
    if ma ~= mb then return ma < mb end
    return a[1] < b[1]
  end)
  local h = 0
  for i = 1, N do h = (h * 31 + data[i][2]) % 1000000007 end
  return h
end)

File: bench/benches/sortby.rs
include!("../lib/bench.rs");
// sortby: N (key, idx) records ordered by a user comparator on a derived key (key%4096, then key);
// checksum = rolling hash of the idx sequence.
const N: usize = 5000;
fn main() {
    bench("sortby", || {
        let mut x: i64 = 1;
        let mut data: Vec<(i64, i64)> = Vec::with_capacity(N);
        for i in 0..N {
            x = (16807 * x) % 2147483647;
            data.push((x, i as i64));
        }
        data.sort_by(|a, b| (a.0 % 4096, a.0).cmp(&(b.0 % 4096, b.0)));
        let mut h: i64 = 0;
        for &(_, i) in &data {
            h = (h * 31 + i) % 1000000007;
        }
        h
    });
}
